### PR TITLE
Figma integration

### PR DESCRIPTION
**Pull request by VanhaverbekeCedric**
**How to test:**

```
npm start -- HbgJuBVOwIOcoZMVnpG01LqA --token [YOURTOKEN]
```

First argument is the Figma file id and the second argument is the authorization token.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,80 +1,54 @@
...
-} = require('./lib/sketch');
-const mappers = require('./lib/mappers');
-const { prettyJSON } = require('./lib/utils');
+const getParser = require('./lib/parser');
+const getMappers = require('./lib/mappers');
+const { prettyJSON} = require('./lib/utils');
 const mapToStyleDictionaryTokens = require('./lib/styleDictionary');
 
-module.exports = (args, flags) => {
+const getExtension = (file) => file.split('.').pop();
```
Comment (thibmaek) [inline index.js]: Should be an util like prettyJSON.

File: index.js [hunk 2]
```diff
@@ -1,80 +1,54 @@
...
-      }
-
-      return response;
-    } catch (err) {
+  const [file] = args
+  const {parser, response} = await getParser(args, flags);
+  const mappers = getMappers(getExtension(file));
+  
+  const mapping = {
+    textStyles: mappers.mapTextStyles(parser.getTextStyles()),
```
Comment (thibmaek) [inline index.js]: Do you think it would be a good solution to do something like:

```js
const { parser, response } = await getParser(args, flags);
const { colors, gradients, shadows, borders } = parser.getTokens();

const mapping = {
  colors: mappers.mapColors(colors),
  ...
};
```

File: lib/figma.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+const Figma = require('figma-js');
+
+module.exports =  {
+    /**
+     * Returns a client to use the Figma API
+     * @param {string} token user token 
+     * @return {Object} 
+     */
+    getClient: (token) => Figma.Client({personalAccessToken: token}),
```
Comment (thibmaek) [inline lib/figma.js]: Just name your arg `personalAccessToken` here and do `Client({ personalAccessToken })`

File: lib/figma.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+     * Helper function for flattenChildren
+     * @param {Object} element a parent element
+     * @param {Arrat<Object>} resultArray the array that will be returned at the end
+     */
+    flattenChildrenHelper: (element, resultArray) => {
+        if(element.children){
+            element.children.forEach(child => {
+                const clone = Object.assign({}, child);
+                resultArray.push(clone);
+                module.exports.flattenChildrenHelper(child, resultArray);
```
Comment (thibmaek) [inline lib/figma.js]: can you not use this? or create this object first and export it later?

File: lib/figma.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    },
+
+    /**
+     * Finds certain elements containing a keyword 
+     * @param {string} keyword 
+     * @param {Array<Object>} components
+     * @return {Array<Object>}
+     */
+    findIds: (keyword, components) => components
```
Comment (thibmaek) [inline lib/figma.js]: nice

File: lib/figma.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    getClient: (token) => Figma.Client({personalAccessToken: token}),
+
+    /**
+     * Executes the API call by giving a valid fileId
+     * @param {Object} client Figma client from figma-js package
+     * @param {string} fileId id from Figma file
+     * @return {Object}
+     */
+    figma2json: async(client, fileId) => {
+        try {
```
Comment (thibmaek) [inline lib/figma.js]: You could also do this:

```js
try {
  return client.file(fileId);
} catch(err) {
  throw new Error(`Failed getting file ${file}:`, err);
}
```

File: lib/mappers/figma/border.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+const {mapColorKeys} = require('../../utils');
+const formatColor = require('../sketch/color');
```
Comment (thibmaek) [inline lib/mappers/figma/border.js:4]: if it is not unique to sketch (hence being reused here) it should move to utils

File: lib/mappers/figma/border.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    const info = border.children[0];
+    return {
+        width: info.strokeWeight,
+        type: info.strokeAlign,
+        radius: border.name.toLowerCase().includes('gradient') 
+        ? info.rectangleCornerRadii || [0]
+        : info.cornerRadius || 0
+    }
+};
+exports.getBorderInfo = getBorderInfo;
```
Comment (thibmaek) [inline lib/mappers/figma/border.js]: why is this exported?

File: lib/mappers/figma/border.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+const {mapColorKeys} = require('../../utils');
+const formatColor = require('../sketch/color');
+const {tokenize} = require ('../../utils');
+const mapGradient = require('./gradient');
+
+const getGradientPropertyLocation = (border) =>  border.children[0].strokes[0];
+exports.getGradientPropertyLocation = getGradientPropertyLocation;
+
+const getBorderInfo = (border) => {
+    const info = border.children[0];
```
Comment (thibmaek) [inline lib/mappers/figma/border.js]: `const [info] = border.children;`

File: lib/mappers/figma/border.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    return {
+        width: info.strokeWeight,
+        type: info.strokeAlign,
+        radius: border.name.toLowerCase().includes('gradient') 
+        ? info.rectangleCornerRadii || [0]
```
Comment (thibmaek) [inline lib/mappers/figma/border.js]: hmm having `number[]` here implies that it should also be available like this in Sketch in this format. Output for Sketch & Figma should be as 1:1 as possible

File: lib/mappers/figma/border.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        gradient : mapGradient(getGradientPropertyLocation(border))
+    });
+exports.getGradientInfo = getGradientInfo;
+
+module.exports = (border) => ({
+    id: border.name.split('/').pop(),
+    token: tokenize('border'),
+    ...getBorderInfo(border),
+    ...((border.name.toLowerCase().includes('gradient') 
+        ? getGradientInfo(border)
```
Comment (thibmaek) [inline lib/mappers/figma/border.js]: Do we have gradient borders available in Sketch? If not, remove gradient borders for now and we will add them later as a feature. For now we need to focus on just getting the Sketch output as is ported to Figma

File: package.json [hunk 12]
```diff
@@ -54,6 +54,7 @@
     "camelcase": "^5.0.0",
     "color-convert": "^1.9.3",
     "color-namer": "^1.3.0",
+    "figma-js": "^1.7.0",
```
Comment (thibmaek) [inline package.json:54]: The lockfile should have changed as well when installing this dependency.

File: lib/utils.js [hunk 13]
```diff
@@ -34,5 +34,34 @@ module.exports = {
    * Get the default line height (font size * 1.2) for font styles
    * missing a user-defined line height.
    */
-  calculateDefaultLineHeight: (fontAttrs) => parseFloat((fontAttrs.size * 1.2).toFixed(2))
+  calculateDefaultLineHeight: (fontAttrs) => parseFloat((fontAttrs.size * 1.2).toFixed(2)),
+
+  /**
+   * Maps keys of a color object from g,b,r,a
```
Comment (thibmaek) [inline lib/utils.js]: Please use the official order RGBA i/o GBRA

File: lib/parser/index.js [hunk 14]
```diff
@@ -0,0 +1,71 @@
+const sketch2json = require('sketch2json');
+const {getClient, figma2json, findIds, flattenChildren} = require('../figma');
+const {
+    getPageArrays, getColorsFromArtboard, getGradientsFromArtboard, getShadowsFromArtboard,
+    getBordersFromArtboard
+  } = require('../../lib/sketch');
+
+const getFileContents = require('../reader');
+
+const getFileExtension = (file) => file.split('.').pop();
```
Comment (thibmaek) [inline lib/parser/index.js]: this is being redefined. You used this before above.

File: lib/parser/index.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+exports.getFileExtension = getFileExtension;
+
+module.exports = async (args, flags) => {
+  const file = args[0];
```
Comment (thibmaek) [inline lib/parser/index.js]:
```js
module.exports = async (args, flags) => {
  const [file] = args;

  const isSketch = /.*sketch/.test(file);

  if (isSketch) {
    console.log('Parsing Sketch file...');
  } else {
	console.log('Parsing Figma file...');
  }
};
```

File: lib/parser/index.js [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        module.getFonts = () => response.meta.fonts;
+        module.getVersion = () => ({sketchVersion: response.meta.appVersion});
+
+        return ({parser: module, response});
+      } catch(err){
+        throw new Error(err);
+      }
+  }
+  // file is figmafile
+  const [fileId, token] = args;
```
Comment (thibmaek) [inline lib/parser/index.js]: Token should be supplied as a CLI flag and not an argument. Please take a look at `cli.js` and add a `--token` flag to use. Before parsing anything Figma-related, check if you have that CLI flag, otherwise throw an error.

File: lib/reader/index.js [hunk 17]
```diff
@@ -0,0 +1,3 @@
+const fs = require('fs').promises;
+
+module.exports = async(file) =>  fs.readFile(file);
```
Comment (thibmaek) [inline lib/reader/index.js]: this is just an util

**Review by thibmaek — CHANGES_REQUESTED** [review]
Also needs:

- Some formatting, lots of semicolons missing, unwanted spaces
- A link to the mock Figma file added somewhere.
- Updated README
- Updated Wiki (once this is merged)

File: index.js [hunk 18]
```diff
@@ -1,80 +1,54 @@
...
-        await fs.writeFile(`${flags.outputDir}/hubble-data.json`, prettyJSON(mapping), fsErrorHandler);
-      }
-
-      if (flags.dump) {
-        await fs.writeFile(`${flags.outputDir}/logdump.json`, prettyJSON(response), fsErrorHandler);
-      }
-
-      return response;
-    } catch (err) {
+  const [file] = args
```
Comment (brecht) [inline index.js]: add semicolon

File: lib/mappers/figma/border.js (hunk 8 above)
Comment (VanhaverbekeCedric) [reply lib/mappers/figma/border.js]: For some reason I thought it was necessary to be able to call the functions inside module.exports. My bad

File: lib/mappers/figma/border.js (hunk 11 above)
Comment (VanhaverbekeCedric) [reply lib/mappers/figma/border.js]: Yes:
https://github.com/inthepocket/hubble-scripts/blob/e53acc52d375e90c61aef4ebb927af0086bc1409/__mocks__/sample_output.document.json#L414

File: lib/parser/index.js (hunk 16 above)
Comment (VanhaverbekeCedric) [reply lib/parser/index.js]: I added it, but when I now run `npm start [fileId] --token [token]`, the command fails. 
`node cli.js  [fileId] --token [token]`still works though...

File: lib/mappers/figma/border.js (hunk 10 above)
Comment (VanhaverbekeCedric) [reply lib/mappers/figma/border.js]: Do you mean the [0] on line 15?
When there is no radius, the output is

`"radius": 0`
I'll change it like that.

File: lib/figma.js (hunk 4 above)
Comment (VanhaverbekeCedric) [reply lib/figma.js]: I don't quite understand. You want this to be removed or used in another way?

File: lib/mappers/__tests__/mapper.figma.test.js [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+describe('figma mapper test suite', () => {
+  describe('-> #formatColors:', () => {
+    it('Given figma formatted array of JSON colors, should transform to a nicely parsed array', () => {
+        expect(mapper.mapColors(FIXTURES.COLORS)).toMatchSnapshot();
+    })
+  })
+})
+  
+  describe('-> #formatTextStyles:', () => {
```
Comment (thibmaek) [inline lib/mappers/__tests__/mapper.figma.test.js:16]: shouldn't these describe blocks be in the describe('figma mapper test suite' block?
Then you could move the before hook there as well:

```js
describe('figma mapper test suite', () => {
	beforeAll(() => {
	  mapper = getMapper('');
	  FIXTURES = getFixtures();
	});

	describe('-> #formatColors:', () => {...});
	describe('-> #formatTextStyles:', () => {...});
});
```

File: lib/utils.js [hunk 20]
```diff
@@ -34,5 +40,72 @@ module.exports = {
...
+  formatColor: color => {
+    const { red, green, blue, alpha } = color;
+    const R = module.exports.convertNSColorToRGB(red);
+    const G = module.exports.convertNSColorToRGB(green);
+    const B = module.exports.convertNSColorToRGB(blue);
+
+    const computedColorName = module.exports.getNamedColorFromRGB([R, G, B]);
+
+    return {
+      token: module.exports.tokenize('color'),
```
Comment (thibmaek) [inline lib/utils.js]: I would just do this, referencing the exported module feels icky:

```js
const utils = {
  namedColor => true,
  formatColor => utils.namedColor()
};

module.exports = utils;
```

File: lib/utils.js [hunk 21]
```diff
@@ -34,5 +40,72 @@ module.exports = {
...
    * missing a user-defined line height.
    */
-  calculateDefaultLineHeight: (fontAttrs) => parseFloat((fontAttrs.size * 1.2).toFixed(2))
+  calculateDefaultLineHeight: fontAttrs => parseFloat((fontAttrs.size * 1.2).toFixed(2)),
+
+  /**
+   * Maps keys of a color object from r,g,b,a
+   * to green, blue, red and alpha
+   */
+  mapColorKeys: colors => {
```
Comment (thibmaek) [inline lib/utils.js]: Why is this needed? All the outputs use RGB in short form instead of longform?

File: lib/utils.js [hunk 22]
```diff
@@ -34,5 +40,72 @@ module.exports = {
    * Get the default line height (font size * 1.2) for font styles
    * missing a user-defined line height.
    */
-  calculateDefaultLineHeight: (fontAttrs) => parseFloat((fontAttrs.size * 1.2).toFixed(2))
+  calculateDefaultLineHeight: fontAttrs => parseFloat((fontAttrs.size * 1.2).toFixed(2)),
+
+  /**
+   * Maps keys of a color object from r,g,b,a
+   * to green, blue, red and alpha
```
Comment (thibmaek) [inline lib/utils.js:51]: Why would we need to map colors to longform?
Output should always be rgb in shortform

File: lib/parser/index.js [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    } catch (err) {
+      throw new Error(err);
+    }
+  }
+};
+
+module.exports = (args, flags) => {
+  module.getTokens = () => parseTokens(args, flags);
+  return {
+    parser: module,
```
Comment (thibmaek) [inline lib/parser/index.js]: Is this not better than directly modifying the global module namespace?

```js
module.exports = (args, flags) => {
  return {
    parser: {
      getTokens: () => parseTokens(args, flags)
    }
  }
}
```

File: lib/parser/index.js [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+  getColorsFromArtboard,
+  getGradientsFromArtboard,
+  getShadowsFromArtboard,
+  getBordersFromArtboard,
+} = require('../../lib/sketch');
+const { readFile, uniqueArray } = require('../utils');
+
+const parseTokens = async (args, flags) => {
+  const [file] = args;
+  const isSketch = /.*sketch/.test(file);
```
Comment (thibmaek) [inline lib/parser/index.js]: I think this regex is supposed to be:

```suggestion
  const isSketch = /*.sketch/.test(file);
```

File: lib/parser/index.js [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+  } else {
+    const { token } = flags;
+    if (!token) {
+      throw new Error('Please add a Figma API authorization token');
+    }
+
+    const client = getClient(token);
+    const response = await figma2jsonFromAPI(client, file);
+
+    try {
```
Comment (thibmaek) [inline lib/parser/index.js]: Why would this need to be wrapped in a trycatch?

Seems like a simple return of an object using functions which only derive data and do not perform any async logic/

File: lib/mappers/index.js [hunk 26]
```diff
@@ -1,13 +1,30 @@
...
+const mapTextStyleFigma = require('./figma/textstyle');
+const mapColorFigma = require('./figma/color');
+const mapGradientFigma = require('./figma/gradient');
+const mapShadowFigma = require('./figma/shadow');
+const mapBorderFigma = require('./figma/border');
+
+module.exports = extension => {
+  const ex = extension.toLowerCase();
+
+  module.mapColors = colors =>
```
Comment (thibmaek) [inline lib/mappers/index.js]: I would never reference global namescapes unless applying a hack or workaround. Is there a reason why you would do this?

File: lib/mappers/index.js [hunk 27]
```diff
@@ -1,13 +1,30 @@
...
+const mapBorderSketch = require('./sketch/border');
+
+const mapTextStyleFigma = require('./figma/textstyle');
+const mapColorFigma = require('./figma/color');
+const mapGradientFigma = require('./figma/gradient');
+const mapShadowFigma = require('./figma/shadow');
+const mapBorderFigma = require('./figma/border');
+
+module.exports = extension => {
+  const ex = extension.toLowerCase();
```
Comment (thibmaek) [inline lib/mappers/index.js]: ext would be the common way of abbreviating extension

File: lib/mappers/index.js [hunk 28]
```diff
@@ -1,13 +1,30 @@
...
+const mapColorSketch = require('./sketch/color');
+const mapGradientSketch = require('./sketch/gradient');
+const mapShadowSketch = require('./sketch/shadow');
+const mapBorderSketch = require('./sketch/border');
+
+const mapTextStyleFigma = require('./figma/textstyle');
+const mapColorFigma = require('./figma/color');
+const mapGradientFigma = require('./figma/gradient');
+const mapShadowFigma = require('./figma/shadow');
+const mapBorderFigma = require('./figma/border');
```
Comment (thibmaek) [inline lib/mappers/index.js]: Place a module mapper at `sketch/index.js` and `figma/index.js` and you can do this:

```js
// figma/index.js
module.exports = {
  colors: mapColorFn,
  ...
};

// this file
const mapFigma = require('./figma');

sketch ? mapSketch.colors : mapFigma.colors;

```

File: lib/utils.js (hunk 22 above)
Comment (VanhaverbekeCedric) [reply lib/utils.js]: lol, other way around

File: lib/mappers/figma/shadow.js [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+  const { color } = shadow.children[0].effects[0];
+  return { color: formatColor(mapColorKeys(color)) };
+};
+
+module.exports = shadow => ({
+  id: shadow.name.split('/').pop(),
+  token: tokenize('shadow'),
+  ...getShadowEffectProperties(shadow),
+  ...getShadowColor(shadow),
+  spread: null, // spread not implemented in figma
```
Comment (thibmaek) [inline lib/mappers/figma/shadow.js:24]: spread is prob. the equiv. of blur in Figma.
Best thing to move forward is to compare outputs side by side for Figma & Sketch

Comment (VanhaverbekeCedric) [reply lib/mappers/figma/shadow.js:24]: I'll check it out.
This was what I found about spread: 
https://spectrum.chat/figma/general/spread-value-in-shadow~148e8d26-7704-40e0-8c67-dbcdedfaaea6

File: lib/utils.js (hunk 22 above)
Comment (VanhaverbekeCedric) [reply lib/utils.js:51]: The reasoning behind it is that in this manner I can make use of the formatColor function which expects the full names. Should i rewrite that part?

File: lib/parser/index.js (hunk 24 above)
Comment (VanhaverbekeCedric) [reply lib/parser/index.js]: Don't you mean 
```js
/.*\.sketch/
```

File: lib/mappers/index.js (hunk 26 above)
Comment (VanhaverbekeCedric) [reply lib/mappers/index.js]: Except my lack of native javascript programming, probably not :p

File: lib/utils.js (hunk 22 above)
Comment (thibmaek) [reply lib/utils.js:51]: I would get longform -> shortform. Would not know why you need shortform -> longform.

**Comment by thibmaek** [comment]
Some conflicts in the previous mocks, they can be removed since the new mocks reside under `__mocks__/{sketch, figma}`

File: __mocks__/sketch/sample_output.artboard.json [hunk 30]
```diff
@@ -1087,6 +1087,5 @@
     "Helvetica-Bold",
     "FiraSans-Regular",
     "FiraSans-Book"
-  ],
-  "sketchVersion": "55"
+  ]
```
Comment (thibmaek) [inline __mocks__/sketch/sample_output.artboard.json]: The sketch version is no longer in the output. You can omit this in Figma since everyone always works against latest version of Figma, but for Sketch we still need this since Sketch updates sometimes contain breaking API updates to tokens.

Please include this again

File: cli.js [hunk 31]
```diff
@@ -12,40 +13,46 @@ const cli = meow(`
     --useColorArtboards             Use artboard formatting to export colors instead of using the document colors
     --useGradientArtboards          Use artboard formatting to export gradients instead of using the document gradients
     --useStyleDictionaryOutput, -s  Generate Style Dictionary compatible output instead of the generic design token format.
+    --token, -t                     Authorization token when accessing the Figma API
```
Comment (thibmaek) [inline cli.js]: Just a detail, but in these help examples please make sure the sentences end with a dot.

File: lib/figma.js [hunk 32]
```diff
@@ -0,0 +1,83 @@
+const Figma = require('figma-js');
+
+let elements = [];
+
+/**
+ * Helper function for flattenChildren
+ * @param {Object} element a parent element
+ * @param {Arrat<Object>} resultArray the array that will be returned at the end
```
Comment (thibmaek) [inline lib/figma.js]:
```suggestion
 * @param {Array<Object>} resultArray the array that will be returned at the end
```

File: lib/figma.js [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+/**
+ * Transforms a Figma formatted JSON object
+ * to a standardized JSON object
+ * and initializes the local variable 'elements'
+ * @param {Object} file Figma formatted JSON file
+ * @return {void}
+ */
+const figma2jsonFromJSON = json => {
```
Comment (thibmaek) [inline lib/figma.js]: This is a bit weird. So this is just the mapping to get Figma API -> Tokens? If so, just name it `figmaToTokens`

File: lib/figma.js [hunk 34]
```diff
@@ -0,0 +1,83 @@
+const Figma = require('figma-js');
+
+let elements = [];
```
Comment (thibmaek) [inline lib/figma.js]: This is really needed. I always get nervous when I see these kind of global mutable values inside a module. What makes it so that you can not have a function that just returns elements inside its own scope, by for example composing functions?

File: lib/mappers/__tests__/__fixtures__/mapper.figma.js [hunk 35]
```diff
@@ -0,0 +1,17 @@
+const sampleOutput = require('../../../../__mocks__/figma/sample_dump.json');
+const { figma2jsonFromJSON, findIds } = require('../../../figma');
+const { uniqueArray } = require('../../../utils');
+
+module.exports = () => {
+  figma2jsonFromJSON(sampleOutput);
```
Comment (thibmaek) [inline lib/mappers/__tests__/__fixtures__/mapper.figma.js]: this is weird? What does this function do when you call it without storing its return value? None of the values below rely on what is returned from this function?

File: lib/mappers/__tests__/__fixtures__/mapper.figma.js [hunk 36]
```diff
@@ -0,0 +1,17 @@
...
+  return ({
+    COLORS : findIds('color'),
```
Comment (thibmaek) [inline lib/mappers/__tests__/__fixtures__/mapper.figma.js]: Keeping consistent with the rest of our voice and tone, this sounds more like `findToken` to me.

File: lib/mappers/__tests__/__fixtures__/mapper.sketch.js [hunk 37]
```diff
@@ -12,7 +12,7 @@ exports.DOCUMENT_COLORS = sampleOutput.document.assets.colorAssets
   .map(({ color, _class, name }) => ({ ...color, _class, name }));
 exports.DOCUMENT_GRADIENTS = sampleOutput.document.assets.gradientAssets
   .map(({ gradient, _class }) => ({ ...gradient, _class }));;
-exports.TEXT_STYLES = sampleOutput.document.layerTextStyles;
+exports.TEXT_STYLES = sampleOutput.document.layerTextStyles.objects;
```
Comment (thibmaek) [inline lib/mappers/__tests__/__fixtures__/mapper.sketch.js:15]: wow, did the Sketch API change in the mean time? Make sure that this still yields the same result as it did before in outputs.

Comment (thibmaek) [reply lib/mappers/__tests__/__fixtures__/mapper.sketch.js:15]: This is a red flag to me when I check the diff for sketch/sample_dump.json

> File renamed without changes.

If it didn't change this will prob. be a regression since you are accessing a layer deeper.

File: lib/mappers/__tests__/mapper.figma.test.js [hunk 38]
```diff
@@ -0,0 +1,44 @@
+const getFixtures = require('./__fixtures__/mapper.figma');
+const getMapper = require('..');
+
+let FIXTURES = {};
+let mappers;
+
+describe('figma mapper test suite', () => {
+  beforeAll(() => {
+    mappers = getMapper(false); // false means no sketchfile
```
Comment (thibmaek) [inline lib/mappers/__tests__/mapper.figma.test.js]: What if we later decide to support Adobe XD? You can only provide true/false now.

Deciding how to map should be the responsibility of the getMapper function. You should just pass `file` to the mapper where it can decide what type of file it is.

Comment (thibmaek) [reply lib/mappers/__tests__/mapper.figma.test.js]: Why are these not just globally scoped? Do they need to be reassigned to the mutable values before every test?

Just create 2 consts at the top and reference those? I don't see any reason why reassigning these values is required.

File: lib/mappers/__tests__/mapper.sketch.test.js [hunk 39]
```diff
@@ -1,7 +1,12 @@
...
+const FIXTURES = require('./__fixtures__/mapper.sketch');
 
-const mapper = require('..');
-const formatColor = require('../color');
+const getMapper = require('..');
+const formatColor = require('../sketch/color');
+
+let mappers;
+beforeAll(() => {
+   mappers = getMapper("sketch");
```
Comment (thibmaek) [inline lib/mappers/__tests__/mapper.sketch.test.js]: Same remark here, why is this not just a globally scoped const and is it being reassigned?

File: lib/mappers/figma/color.js [hunk 40]
```diff
@@ -0,0 +1,7 @@
+const { mapColorKeys, formatColor } = require('../../utils');
+
+const getColorProperty = component => component.children[0].fills[0].color;
```
Comment (thibmaek) [inline lib/mappers/figma/color.js]: Why put this in a separate function when you can just provide it as an argument? I see no reuse and no reason to put it in its own function.

File: lib/mappers/figma/gradient.js [hunk 41]
```diff
@@ -0,0 +1,40 @@
...
+const getGradientStops = gradient => {
+  const { gradientStops } = getGradientPropertyLocation(gradient);
+  return {
+    stops:
+      gradientStops !== undefined
+        ? gradientStops.map(stop => ({
+            position: stop.position,
+            color: formatColor(mapColorKeys(stop.color)),
+          }))
+        : null,
```
Comment (thibmaek) [inline lib/mappers/figma/gradient.js]: i would expect stops to always be an array, so an empty one in this case. Verify how we do this in Sketch

File: lib/mappers/figma/gradient.js [hunk 42]
```diff
@@ -0,0 +1,40 @@
...
+  return {
+    from: gradientHandlePositions
+      ? [gradientHandlePositions[0].x, gradientHandlePositions[0].y]
+      : null,
+    to: lastGradient ? [lastGradient.x, lastGradient.y] : null,
+  };
+};
+
+module.exports = gradient => ({
+  name: gradient.name ? gradient.name.split('/').pop() : null,
```
Comment (thibmaek) [inline lib/mappers/figma/gradient.js]: names should never be null, if we have no name this is not a valid gradient and the program should crash

File: lib/utils.js [hunk 43]
```diff
@@ -1,38 +1,131 @@
...
+const colorConvert = require('color-convert');
 const colorNamer = require('color-namer');
+const fs = require('fs').promises;
+
+const isArtboardColor = color => color._class === 'artboardColor';
+exports.isArtboardColor = isArtboardColor;
+exports.isDocumentColor = color =>
+  color._class === 'color' || color._class === 'MSImmutableColorAsset';
+
+const fsErrorHandler = err => {
```
Comment (thibmaek) [inline lib/utils.js]: If you don't export this and reuse it, write it inline. Don't abstract for the reason to abstract.

File: lib/utils.js [hunk 44]
```diff
@@ -1,38 +1,131 @@
...
+  calculateDefaultLineHeight: fontAttrs => parseFloat((fontAttrs.size * 1.2).toFixed(2)),
 
   /**
-   * Formats design tokens to standardised Hubble token names.
-   * These tokens are kebab-cased
+   * Maps keys of a color object from r,g,b,a
+   * to green, blue, red and alpha
    */
-  tokenize: token => token.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase(),
+  mapColorKeys: colors => {
```
Comment (thibmaek) [inline lib/utils.js:53]: You will need to write a unit test for this inside utils.test.js

**Review by thibmaek — CHANGES_REQUESTED** [review]

File: __mocks__/sketch/sample_output.artboard.json (hunk 30 above)
Comment (VanhaverbekeCedric) [reply __mocks__/sketch/sample_output.artboard.json]: It was included again, appareantly I didn't push the updated sample_output

File: lib/mappers/figma/gradient.js (hunk 42 above)
Comment (VanhaverbekeCedric) [reply lib/mappers/figma/gradient.js]: I did this because I wasn't sure how to call the gradient. In Sketch the name of following gradient is `norwayGradient`

```json
{
            "id": "1:60",
            "name": "primitives/border/gradient-with-corners",
            "type": "FRAME",
            "blendMode": "PASS_THROUGH",
            "children": [
              {
                "id": "1:61",
                "name": "Rectangle",
                "type": "RECTANGLE",
                "blendMode": "PASS_THROUGH",
                "absoluteBoundingBox": {
                  "x": 336,
                  "y": 54,
                  "width": 80,
                  "height": 25
                },
                "constraints": {
                  "vertical": "SCALE",
                  "horizontal": "SCALE"
                },
                "fills": [
                  {
                    "blendMode": "NORMAL",
                    "type": "SOLID",
                    "color": {
                      "r": 0.847000002861023,
                      "g": 0.847000002861023,
                      "b": 0.847000002861023,
                      "a": 1
                    }
                  }
                ],
                "strokes": [
                  {
                    "blendMode": "NORMAL",
                    "type": "GRADIENT_LINEAR",
                    "gradientHandlePositions": [
                      {
                        "x": 1,
                        "y": 1
                      },
                      {
                        "x": 1,
                        "y": 0
                      },
                      {
                        "x": 1.5,
                        "y": 0.9999999999999999
                      }
                    ],
                    "gradientStops": [
                      {
                        "color": {
                          "r": 0.5921568870544434,
                          "g": 0.7137255072593689,
                          "b": 0.5647059082984924,
                          "a": 1
                        },
                        "position": 0
                      },
                      {
                        "color": {
                          "r": 0.3333333432674408,
                          "g": 0.5686274766921997,
                          "b": 0.3450980484485626,
                          "a": 1
                        },
                        "position": 0.5038507580757141
                      },
                      {
                        "color": {
                          "r": 0.13333334028720856,
                          "g": 0.4901960790157318,
                          "b": 0.24313725531101227,
                          "a": 1
                        },
                        "position": 1
                      }
                    ]
                  }
                ],
                "strokeWeight": 1,
                "strokeAlign": "OUTSIDE",
                "effects": [],
                "rectangleCornerRadii": [
                  8,
                  12.5,
                  12.5,
                  12.5
                ]
              }
            ],
            "absoluteBoundingBox": {
              "x": 325,
              "y": 45,
              "width": 102,
              "height": 44
            },
            "constraints": {
              "vertical": "SCALE",
              "horizontal": "SCALE"
            },
            "clipsContent": true,
            "background": [
              {
                "blendMode": "NORMAL",
                "type": "SOLID",
                "color": {
                  "r": 1,
                  "g": 1,
                  "b": 1,
                  "a": 1
                }
              }
            ],
            "backgroundColor": {
              "r": 1,
              "g": 1,
              "b": 1,
              "a": 1
            },
            "layoutGrids": [],
            "effects": []
          }
```

File: lib/mappers/figma/gradient.js (hunk 41 above)
Comment (VanhaverbekeCedric) [reply lib/mappers/figma/gradient.js]: This was my bad, normally all gradients should have stops. In Sketch there is no case defined when there are no stops.

File: lib/figma.js (hunk 34 above)
Comment (VanhaverbekeCedric) [reply lib/figma.js]: Something like this? :
```
const data = () => {
  let elements = [];
  return {
    setElements : (element) => {elements = flattenChildren(element)},
    getElements : () => elements
  }
}
```

Comment (VanhaverbekeCedric) [reply lib/figma.js]: I changed it to this:
```
const data = (() => {
  let elements;
  return {
    setElements : (element) => {elements = flattenChildren(element)},
    getElements : () => elements
  };
})();
```

File: index.js [hunk 45]
```diff
@@ -1,80 +1,53 @@
...
+    colors,
+    gradients,
+    shadows,
+    borders,
+    fonts,
+    version,
+    response,
+  } = await parser.getTokens();
+
+  const mappers = getMappers(isSketch(file));
```
Comment (thibmaek) [inline index.js]: Doesn't the parser already check for Sketch/Figma?

Maybe the parser should return a property `fileType: "SKETCH" | "FIGMA"` and you just pass this to `getMappers(fileType)`?

Comment (thibmaek) [reply index.js]: Now we are just providing `getMappers(true | false)` but which one signifies Sketch or Figma? There is no outside knowledge if you look at this function outside of this file.

File: lib/__tests__/utils.test.js [hunk 46]
```diff
@@ -16,4 +16,16 @@ describe('[utils.js] test suite', () => {
       expect(uniqueArray(arrayWithDupes)).toEqual([1, 2, `a`, `b`]);
     });
   });
+
+  describe('-> #mapColorKeys', () => {
+    describe('should output an object with full name keys', () => {
+      const color = {
+        r: "",
```
Comment (thibmaek) [inline lib/__tests__/utils.test.js]: I would fill in these values.

**Comment by thibmaek** [comment]
Make sure the conflicts are fixed so we can compare the output side by side. The snapshot should be regenerated, the previous JSON mocks deleted and the Sketch file fixed.

**Comment by thibmaek** [comment]
I removed the usage of `fs.promises` since we are still using the node@8 runtime for the moment and the stage is experimental: https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fs_promises_api

**Review by thibmaek — APPROVED** [review]
